src/SAML2/XML/samlp/Status.php

```diff
@@ -66,7 +67,7 @@ public function getStatusCode(): StatusCode
      *
      * @throws \InvalidArgumentException if assertions are false
      */
-    public function setStatusCode(StatusCode $statusCode): void
+    private function setStatusCode(StatusCode $statusCode): void
```

Comment:
I'm not completely convinced we should declare all setters as private. In some cases, I think it would be desirable to extend an existing class to implement improved functionality, and then access to those methods would be needed. Do we already have any classes under `SAML2\XML` that extend others?

Reply:
We extend from SignedElementHelper, RoleDescriptor, SSODescriptorType, EndpointType, NameIDType and Attribute..

I can only imagine doing so if you want to support some almost-but-not-really-SAML spec.. Strictly speaking the classes should be final, but I guess that's pretty harsh..

Reply:
I actually like the approach to make classes final better than making setters private. In the end, the standard has its own extension points, where you should add your functionality, whatever that shall be. We should follow the same principle, and the rest of the classes representing elements defined by SAML2 should probably be final as you are saying, unless we have a reason not to make them so (e.g. we're extending them somewhere, but in that case we're talking about types, rather than actual elements, so for those it wouldn't make sense to mark them as final.

I'm wondering if we should use traits instead of an abstract class that we are extending...

Reply:
If we make them public again, we have to add a lot of protecting assertions to toXML again to avoid illegal situations.. I thought we had agreed on the immutable objects..
I'd say final+private and/or non-final+protected..

Reply:
No, no, I didn't mean making them public, but protected. But of course that would only make sense for non-final classes. Also, provided that the setters are protected, I'm fine with the properties being private, since any class potentially extending can still get and set those values.